code/game/objects/structures/janicart.dm

```diff
@@ -1,54 +1,98 @@
-			add_fingerprint(user)
-			user.visible_message("\blue [user] has fixed some of the dents on [src].", "\blue You fix some of the dents on \the [src]")
-			health += 20
-			HealthCheck()
-		else
-			user << "Need more welding fuel!"
-			return
-	if(istype(W, /obj/item/key))
-		user << "Hold [W] in one of your hands while you drive this [name]."
+/obj/structure/stool/bed/chair/janicart/Move(a, b, flag)
+	..()
+	if(floorbuffer)
+		var/turf/tile = loc
+		if(isturf(tile))
+			tile.clean_blood()
+			if (istype(tile, /turf/simulated/floor))
+				var/turf/simulated/floor/F = tile
+				F.dirt = 0
+			for(var/A in tile)
+				if(istype(A, /obj/effect))
+					if(is_cleanable(A))
+						qdel(A)
+
+/obj/structure/stool/bed/chair/janicart/examine(mob/user)
+	..()
+	if(floorbuffer)
+		user << "It has been upgraded with a floor buffer."
 
 
-/obj/structure/stool/bed/chair/cart/user_buckle_mob(mob/living/M, mob/user)
+/obj/structure/stool/bed/chair/janicart/attackby(obj/item/I, mob/user, params)
+	if(istype(I, keytype))
+		user << "Hold [I] in one of your hands while you drive this [callme]."
+	else if(istype(I, /obj/item/weapon/storage/bag/trash))
+		if(keytype == /obj/item/key/janitor)
+			if(!user.drop_item())
+				return
+			user << "<span class='notice'>You hook the trashbag onto the [callme].</span>"
+			I.loc = src
+			mybag = I
+	else if(istype(I, /obj/item/janiupgrade))
+		if(keytype == /obj/item/key/janitor)
+			floorbuffer = 1
+			qdel(I)
+			user << "<span class='notice'>You upgrade the [callme] with the floor buffer.</span>"
+	update_icon()
+
+/obj/structure/stool/bed/chair/janicart/update_icon()
+	overlays.Cut()
+	if(mybag)
+		overlays += "cart_garbage"
+	if(floorbuffer)
+		overlays += "cart_buffer"
+
+/obj/structure/stool/bed/chair/janicart/attack_hand(mob/user)
+	if(mybag)
+		mybag.loc = get_turf(user)
+		user.put_in_hands(mybag)
+		mybag = null
+		update_icon()
+	else
+		..()
+
+
+/obj/structure/stool/bed/chair/janicart/relaymove(mob/user, direction)
+	if(user.stat || user.stunned || user.weakened || user.paralysis)
+		unbuckle_mob()
+	if(istype(user.l_hand, keytype) || istype(user.r_hand, keytype))
+		if(!Process_Spacemove(direction) || !has_gravity(src.loc) || move_delay || !isturf(loc))
+			return
+		step(src, direction)
+		update_mob()
+		handle_rotation()
+		if(istype(src.loc, /turf/simulated))
+			var/turf/simulated/T = src.loc
+			if(T.wet == 2)	//Lube! Fall off!
+				playsound(src, 'sound/misc/slip.ogg', 50, 1, -3)
+				buckled_mob.Stun(7)
+				buckled_mob.Weaken(7)
+				unbuckle_mob()
+				step(src, dir)
+		move_delay = 1
+		spawn(2)
+			move_delay = 0
+	else
+		user << "<span class='notice'>You'll need the keys in one of your hands to drive this [callme].</span>"
+
+/obj/structure/stool/bed/chair/janicart/user_buckle_mob(mob/living/M, mob/user)
 	if(user.incapacitated()) //user can't move the mob on the janicart's turf if incapacitated
 		return
 	for(var/atom/movable/A in get_turf(src)) //we check for obstacles on the turf.

```

Comment:
This entire check could be a bit better; Perhaps checking CanPass instead of a flat density check.